### PR TITLE
Add external-dns for AWS DNS setup and Terraform 0.12 syntax cleanup

### PR DESCRIPTION
**Pull request by satadruroy**
- Add external-dns for AWS DNS.
- Clean up TF 0.12 syntax

### CODE REVIEW

File: aks/k8sresources.tf [hunk 1]
```diff
@@ -1,39 +1,11 @@
...
-        kind = "ClusterRole"
-        name = "cluster-admin"
-    }
-    subject {
-        kind = "ServiceAccount"
-        name = "tiller"
-        namespace = "kube-system"
-    }
-    depends_on = ["kubernetes_service_account.tiller"]
+  version = "~> 1.10.0"
```
Comment (bikramnehra) [inline aks/k8sresources.tf:3]: is this the k8s version? its not clear what version will be used.

File: aks/nginx-ingress.tf [hunk 2]
```diff
@@ -26,5 +26,5 @@ resource "helm_release" "nginx_ingress" {
       value = "true"
   }
 
-  depends_on = ["helm_release.external-dns"]
+  depends_on = [helm_release.external-dns]
```
Comment (bikramnehra) [inline aks/nginx-ingress.tf:29]: can we add a newline at the end of this file.

File: eks/modules.tf [hunk 3]
```diff
@@ -5,20 +5,25 @@ module "network" {
...
+    generated-cluster-name  = module.network.generated-cluster-name
+    vpc-id      = module.network.vpc_id
+    app_subnet_ids = module.network.app_subnet_ids
+    workstation_cidr_block = var.workstation_cidr_block
+    aws-network-dependency-id = module.network.aws-network-dependency
+    keypair_name  = var.keypair_name
+    eks_version   = var.eks_version
+    cluster_labels = var.cluster_labels
+    instance_type  = var.instance_type
+    hosted_zone_policy_arn=var.hosted_zone_policy_arn
```
Comment (bikramnehra) [inline eks/modules.tf:17]: we must add proper spacing here.

File: eks/modules/eks/cluster-iam.tf [hunk 4]
```diff
@@ -11,17 +11,17 @@ data "aws_iam_policy_document" "cluster-role-policy" {
   }
 }
 resource "aws_iam_role" "aws-cluster" {
-  name = "${var.generated-cluster-name}-iam-role"
-  assume_role_policy = "${data.aws_iam_policy_document.cluster-role-policy.json}"
-  
+  name               = "${var.generated-cluster-name}-iam-role"
```
Comment (bikramnehra) [inline eks/modules/eks/cluster-iam.tf:14]: This extra tab must be removed.

File: eks/modules/eks/cluster-security-group.tf [hunk 5]
```diff
@@ -19,19 +19,18 @@ resource "aws_security_group_rule" "aws-cluster-ingress-node-https" {
...
-  cidr_blocks       = ["${var.workstation_cidr_block}"]
+  cidr_blocks       = [var.workstation_cidr_block]
   description       = "Allow workstation to communicate with the cluster API Server"
   from_port         = 443
   protocol          = "tcp"
-  security_group_id = "${aws_security_group.aws-cluster.id}"
+  security_group_id = aws_security_group.aws-cluster.id
   to_port           = 443
   type              = "ingress"
 }
```
Comment (bikramnehra) [inline eks/modules/eks/cluster-security-group.tf:36]: again, we must add extra line here.

File: eks/modules/eks/worker-iam.tf [hunk 6]
```diff
@@ -12,30 +12,40 @@ data "aws_iam_policy_document" "worker-role-policy" {
...
+  role       = aws_iam_role.aws-node.name
 }
 
 resource "aws_iam_role_policy_attachment" "aws-node-AmazonEC2ContainerRegistryReadOnly" {
   policy_arn = "arn:aws:iam::aws:policy/AmazonEC2ContainerRegistryReadOnly"
-  role       = "${aws_iam_role.aws-node.name}"
+  role       = aws_iam_role.aws-node.name
 }
 
 #resource "aws_iam_role_policy_attachment" "worker-ext-dns-policy" {
```
Comment (bikramnehra) [inline eks/modules/eks/worker-iam.tf:33]: should we get rid of the commented stuff if its not needed?

File: eks/modules/eks/worker-iam.tf [hunk 7]
```diff
@@ -12,30 +12,40 @@ data "aws_iam_policy_document" "worker-role-policy" {
...
 #  policy_arn = "${aws_iam_policy.ext-dns-policy.arn}"
 #  role       = "${aws_iam_role.aws-node.name}"
 #}
 
+resource "aws_iam_role_policy_attachment" "worker-ext-dns-policy" {
+  policy_arn = var.hosted_zone_policy_arn
+  role       = aws_iam_role.aws-node.name
+}
+
+#resource "aws_iam_role_policy_attachment" "aws-node-AmazonRoute53FullAccess" {
```
Comment (bikramnehra) [inline eks/modules/eks/worker-iam.tf:43]: same for this.

File: eks/modules/services/aws_auth_cm.tf [hunk 8]
```diff
@@ -19,5 +19,5 @@ resource "kubernetes_config_map" "aws_auth" {
     - system:nodes
 ROLES
 }
-depends_on = ["null_resource.force-eks-dependency"]
+depends_on = [null_resource.force-eks-dependency]
 }
```
Comment (bikramnehra) [inline eks/modules/services/aws_auth_cm.tf:23]: newline needed(last newline comment but assumes will be fixed for rest of the files).

File: gke/main.tf [hunk 9]
```diff
@@ -1,24 +1,24 @@
...
 
   # We can't create a cluster with no node pool defined, but we want to only use
   # separately managed node pools. So we create the smallest possible default
   # node pool and immediately delete it.
   remove_default_node_pool = true
 
   initial_node_count = 1
-  min_master_version = "${var.k8s_version}"
-  node_version = "${var.k8s_version}"
+ # min_master_version = "${var.k8s_version}"
```
Comment (bikramnehra) [inline gke/main.tf:18]: is this needed anymore? if no we should remove it.

File: aks/k8sresources.tf (hunk 1 above)
Comment (satadruroy) [reply aks/k8sresources.tf:3]: No, this is the Terraform kubernetes provider version.

File: gke/main.tf (hunk 9 above)
Comment (satadruroy) [reply gke/main.tf:19]: Removed.

File: eks/modules/eks/worker-iam.tf (hunk 7 above)
Comment (satadruroy) [reply eks/modules/eks/worker-iam.tf:43]: Removed.

File: eks/modules/eks/worker-iam.tf (hunk 6 above)
Comment (satadruroy) [reply eks/modules/eks/worker-iam.tf:33]: Removed.

File: eks/modules/eks/cluster-security-group.tf (hunk 5 above)
Comment (satadruroy) [reply eks/modules/eks/cluster-security-group.tf:36]: Extra line?

File: eks/modules/eks/cluster-iam.tf (hunk 4 above)
Comment (satadruroy) [reply eks/modules/eks/cluster-iam.tf:14]: Removed.

File: eks/modules.tf (hunk 3 above)
Comment (satadruroy) [reply eks/modules.tf:17]: Done.

File: aks/nginx-ingress.tf (hunk 2 above)
Comment (satadruroy) [reply aks/nginx-ingress.tf:29]: Done.

File: eks/modules/services/aws_auth_cm.tf (hunk 8 above)
Comment (satadruroy) [reply eks/modules/services/aws_auth_cm.tf:23]: > Seems like the changes are not just limited to AWS DNS setup, can we update the description or break down the changes in multiple PRs? Also, can we look at merge conflicts.

Resolved merge conflicts.

**Pull request merged**